toolchain/base/value_store_chunk.h

```diff
@@ -0,0 +1,177 @@
+
+#ifndef CARBON_TOOLCHAIN_BASE_VALUE_STORE_CHUNK_H_
+#define CARBON_TOOLCHAIN_BASE_VALUE_STORE_CHUNK_H_
+
+#include <bit>
+#include <cstddef>
+#include <memory>
+#include <type_traits>
+#include <utility>
+
+#include "common/check.h"
+#include "llvm/ADT/StringRef.h"
+#include "toolchain/base/mem_usage.h"
+
+namespace Carbon::Internal {
+
+// The max size of each chunk allocation for ValueStore. This is based on TLB
+// page sizes for the target platform.
+//
+// See https://docs.kernel.org/admin-guide/mm/hugetlbpage.html
+template <class IdT>
+  requires requires { typename IdT::ValueType; }
+static constexpr auto PlatformChunkMaxAllocationBytes() -> int {
+#if !defined(NDEBUG) || defined(FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION)
+  // Use a small size in debug/fuzzer builds to ensure multiple chunks get used.
+  return sizeof(typename IdT::ValueType) * 5;
+#else
+  // TODO: Should ia64 use 1M or 4M? Should Windows and Mac use different sizes?
+
+  // Linux x64 uses 2M, as x64 CPUs support 4K and 2M page sizes.
+  return 2 * 1024 * 1024;
+#endif
+}
+
+// The number of elements stored in each chunk allocation.
+//
+// The number must be a power of two so that that there are no unused values in
+// bits indexing into the allocation.
+template <class IdT>
+  requires requires { typename IdT::ValueType; }
+static constexpr auto PlatformChunkCapacitySize() -> int {
+  constexpr auto MaxElements =
+      PlatformChunkMaxAllocationBytes<IdT>() / sizeof(typename IdT::ValueType);
+  constexpr auto Pow2MaxElements = std::bit_ceil(MaxElements);
+  if (Pow2MaxElements > MaxElements) {
+    return Pow2MaxElements / 2;
+  }
+  return Pow2MaxElements;
+}
+
+// The number of bits needed to index each element in a chunk allocation.
+template <class IdT>
+  requires requires { typename IdT::ValueType; }
+static constexpr auto PlatformChunkCapacityBits() -> int {
+  static_assert(PlatformChunkCapacitySize<IdT>() > 1);
+  int bits = 0;
+  for (auto size = PlatformChunkCapacitySize<IdT>(); size > 1; size /= 2) {
+    ++bits;
+  }
+  return bits;
+}
+
+// Converts an id into an index into the set of chunks, and an offset into that
+// specific chunk.
+template <typename IdT>
+  requires requires { typename IdT::ValueType; }
+static constexpr auto IdToChunkIndices(IdT id) -> std::pair<int, int> {
+  constexpr auto LowBits = PlatformChunkCapacityBits<IdT>();
+
+  // Verify there are no unused bits when indexing up to the
+  // PlatformChunkCapacitySize(). This ensures that ids are contiguous values
+  // from 0, as if the values were all stored in a single array, and allows
+  // using the ids to index into other arrays.
+  static_assert((1 << LowBits) == PlatformChunkCapacitySize<IdT>());
+  // Simple check to make sure nothing went wildly wrong with the
+  // PlatformChunkCapacitySize, and we have some room for a chunk index, and
+  // that shifting by the number of bits won't be UB in an int32_t.
+  static_assert(LowBits < 24);
```

Comment:
Similar to the above for verification in the calculation rather than use; I'll also suggest here though since maybe we'll end up with slightly larger allocation sizes:

```suggestion
  static_assert((1 << LowBits) <= PlatformChunkMaxAllocationBytes());
```

Reply:
That ends up just testing that `PlatformChunkCapacity < PlatformChunkMaxAllocationBytes` (when combined with the equality check above) which is not what I intended here really. How about I use 30 then. I want to make sure the LowBits isn't taking up all of (or more than) the bits in int32. There's no way we would even accept only 2 bits for indexing the chunks (only allowing 4 chunks), right?